# Regresión Lineal con librerías nativas de Python (`csv` y `math`)

A continuación, encontrarás la guía paso a paso y el script en Python para automatizar el cálculo de la regresión lineal simple. El código está dividido en bloques independientes y utiliza exclusivamente las librerías nativas de Python (`csv` y `math`), ejecutando de forma exacta las mismas fórmulas matemáticas que desarrollamos a mano.

## Caso de Estudio: Predicción de Calificaciones basadas en Horas de Estudio

Este caso analiza el impacto de las horas de estudio ($X$) sobre la calificación final ($Y$) obtenida por un grupo de estudiantes en un examen. El objetivo es construir un modelo matemático predictivo capaz de cuantificar esta relación a través de una línea recta de regresión, evaluando además su margen de error y su nivel de precisión.

## Documentación del Dataset

El modelo consume un archivo llamado `datos_estudio.csv`. Este archivo contiene dos columnas numéricas:

* `horas_estudio`: Variable independiente ($X$). Representa el tiempo total invertido en horas por el estudiante.
* `calificacion`: Variable dependiente ($Y$). Representa el puntaje obtenido en el examen (en una escala de 0 a 10).

## Preparación del Archivo de Datos

Para ejecutar el script, primero debes asegurarte de crear el archivo de texto plano llamado `datos_estudio.csv` en la misma carpeta donde guardarás el código Python, con el siguiente contenido exacto:

```text
horas_estudio,calificacion
1,2
2,5
4,5
5,8
```

## Bloque 1: Lectura y Extracción de Datos

**Explicación:**

En este primer bloque utilizamos el módulo nativo `csv` para abrir el archivo y extraer la información sin depender de librerías externas. Recorremos cada fila saltándonos el encabezado, convertimos los textos en números flotantes (float) y los almacenamos en dos listas independientes: `X` (horas de estudio) y `Y` (calificaciones). También calculamos `n`, que representa el número total de observaciones mediante la función `len()`.

In [2]:
import csv

# Inicializar listas para las variables
X = []
Y = []

# Leer el archivo CSV local
with open('datos_estudio.csv', mode='r', encoding='utf-8') as archivo:
    lector_csv = csv.reader(archivo)
    next(lector_csv)  # Omitir el encabezado (horas_estudio, calificacion)
    
    for fila in lector_csv:
        X.append(float(fila[0]))
        Y.append(float(fila[1]))

# Número total de datos (n)
n = len(X)

print(f"Datos cargados correctamente.")
print(f"X (Horas de estudio): {X}")
print(f"Y (Calificaciones):  {Y}")
print(f"Número de muestras (n): {n}")

Datos cargados correctamente.
X (Horas de estudio): [1.0, 2.0, 4.0, 5.0]
Y (Calificaciones):  [2.0, 5.0, 5.0, 8.0]
Número de muestras (n): 4


## Bloque 2: Cálculo de Promedios ($\bar{x}$ y $\bar{y}$)

**Explicación:**

Para encontrar la línea de ajuste, calculamos el "centro de gravedad" de los datos utilizando las fórmulas de la media aritmética:

$$\bar{x} = \frac{\sum x_i}{n} \quad \text{y} \quad \bar{y} = \frac{\sum y_i}{n}$$

Usamos la función nativa `sum()` de Python dividida entre `n` para obtener los promedios de ambas listas.

In [3]:
# Calcular los promedios de X y Y
promedio_x = sum(X) / n
promedio_y = sum(Y) / n

print(f"sumaX: {sum(X)}")
print(f"sumaY: {sum(Y)}")
print(f"Promedio de Horas de Estudio (x̄): {promedio_x}")
print(f"Promedio de Calificaciones (ȳ): {promedio_y}")

sumaX: 12.0
sumaY: 20.0
Promedio de Horas de Estudio (x̄): 3.0
Promedio de Calificaciones (ȳ): 5.0


## Bloque 3: Cálculo de la Pendiente ($b_1$) e Intercepto ($b_0$)

**Explicación:**

Este bloque implementa las fórmulas centrales de los Mínimos Cuadrados Ordinarios (MCO):

$$b_1 = \frac{\sum (x_i - \bar{x})(y_i - \bar{y})}{\sum (x_i - \bar{x})^2} \quad \text{y} \quad b_0 = \bar{y} - b_1 \bar{x}$$

A través de un ciclo `for`, calculamos las desviaciones de cada punto respecto a su promedio. Multiplicamos las desviaciones cruzadas para el numerador y elevamos al cuadrado las de $X$ para el denominador, sumándolas de forma acumulativa para resolver la división final.

In [4]:
# Inicializar acumuladores para el numerador y denominador de b1
numerador_b1 = 0
denominador_b1 = 0

# Ciclo para calcular las sumatorias paso a paso
for i in range(n):
    desviacion_x = X[i] - promedio_x
    desviacion_y = Y[i] - promedio_y
    
    numerador_b1 += desviacion_x * desviacion_y
    denominador_b1 += desviacion_x ** 2

# Calcular Coeficientes
b1 = numerador_b1 / denominador_b1
b0 = promedio_y - (b1 * promedio_x)

print(f"Pendiente (b1): {b1:.2f}")
print(f"Intercepto (b0): {b0:.2f}")
print(f"Ecuación estimada: ŷ = {b0:.2f} + {b1:.2f}x")

Pendiente (b1): 1.20
Intercepto (b0): 1.40
Ecuación estimada: ŷ = 1.40 + 1.20x


## Bloque 4: Cálculo del Error Estándar ($S_e$) y Determinación ($R^2$)

**Explicación:**

Calculamos el margen de error promedio del modelo (Error Estándar de la Estimación, $S_e$) y el Coeficiente de Determinación ($R^2$) aplicando sus respectivas fórmulas estructurales:

$$S_e = \sqrt{\frac{\sum (y_i - \hat{y}_i)^2}{n - 2}} \quad \text{e} \quad R^2 = 1 - \frac{\text{SCE}}{\text{STC}}$$

Donde $\text{SCE} = \sum (y_i - \hat{y}_i)^2$ representa la Suma de Cuadrados del Error (variabilidad no explicada) y $\text{STC} = \sum (y_i - \bar{y})^2$ es la Suma Total de Cuadrados (variabilidad total de la muestra). Importamos el módulo math únicamente para usar la raíz cuadrada (`math.sqrt`).

In [5]:
import math

# Inicializar acumuladores de variabilidad
sce = 0  # Suma de Cuadrados del Error
stc = 0  # Suma Total de Cuadrados

for i in range(n):
    # Calcular la predicción (ŷ) para el valor actual de X
    y_estimado = b0 + (b1 * X[i])
    
    # Acumular error residual al cuadrado
    sce += (Y[i] - y_estimado) ** 2
    # Acumular variabilidad total al cuadrado
    stc += (Y[i] - promedio_y) ** 2

# Calcular Error Estándar (Se) con n-2 grados de libertad
se = math.sqrt(sce / (n - 2))

# Calcular Coeficiente de Determinación (R²)
r_cuadrado = 1 - (sce / stc)

# Calcular Coeficiente de Correlación de Pearson (r) como la raíz de R²
r_pearson = math.sqrt(r_cuadrado)

print(f"Suma de Cuadrados del Error (SCE): {sce:.2f}")
print(f"Suma Total de Cuadrados (STC): {stc:.2f}")
print(f"Error Estándar de Estimación (Se): {se:.2f}")
print(f"Coeficiente de Determinación (R²): {r_cuadrado:.2f} ({r_cuadrado*100:.1f}%)")
print(f"Coeficiente de Correlación de Pearson (r): {r_pearson:.3f}")

Suma de Cuadrados del Error (SCE): 3.60
Suma Total de Cuadrados (STC): 18.00
Error Estándar de Estimación (Se): 1.34
Coeficiente de Determinación (R²): 0.80 (80.0%)
Coeficiente de Correlación de Pearson (r): 0.894


## Bloque 5: Módulo de Predicción de Nuevos Valores

**Explicación:**

Este último bloque demuestra la utilidad práctica de la ecuación $\hat{y} = b_0 + b_1 x$. Definimos una función que recibe cualquier número de horas de estudio ingresado por el usuario, calcula de forma automática la calificación pronosticada y la imprime en pantalla de manera limpia.

In [6]:
def predecir_calificacion(horas):
    calificacion_predicha = b0 + (b1 * horas)
    return calificacion_predicha

# Ejemplo de prueba con un valor nuevo: 3 horas de estudio
horas_prueba = 3.0
resultado = predecir_calificacion(horas_prueba)

print(f"--- MÓDULO DE PREDICCIÓN ---")
print(f"Si un estudiante dedica {horas_prueba} horas a estudiar:")
print(f"La calificación estimada por el modelo es de: {resultado:.2f} puntos")

--- MÓDULO DE PREDICCIÓN ---
Si un estudiante dedica 3.0 horas a estudiar:
La calificación estimada por el modelo es de: 5.00 puntos


## Tabla Resumen del Modelo de Regresión Lineal

| Indicador Técnico | Variable | Valor en Script | Significado e Interpretación Práctica |
|---|---|---|---|
| **Pendiente** | $b_1$ | `1.20` | Por cada hora adicional de estudio, la calificación aumenta **1.2 puntos**. |
| **Intercepto** | $b_0$ | `1.40` | Un estudiante con 0 horas de estudio tiene una calificación base estimada de **1.4 puntos**. |
| **Error Estándar** | $S_e$ | `1.34` | Las predicciones del modelo fallan, en promedio, por **$\pm 1.34$** puntos frente a las notas reales. |
| **Determinación** | $R^2$ | `0.80` | El **80% de los cambios** en las calificaciones se explican por el tiempo de estudio. |
| **Correlación** | $r$ | `0.894` | Existe una relación lineal **positiva y muy fuerte** entre ambas variables. |

## Resumen Técnico de todo lo aplicado

1. **Extracción limpia:** El script procesó un archivo estructurado csv transformando cadenas de texto en arreglos numéricos flotantes aptos para cálculos matemáticos.
2. **Ajuste por Mínimos Cuadrados:** Se evitó el uso de librerías avanzadas como scikit-learn o statsmodels para evidenciar paso a paso la convergencia matemática de las fórmulas manuales dentro de ciclos lógicos nativos.
3. **Evaluación de Variabilidad:** Se descompuso la varianza total de los datos ($STC = 18.0$) frente a los residuos del modelo ($SCE = 3.6$). Esto demostró estadísticamente que el modelo reduce la incertidumbre inicial en un 80%, consolidando un algoritmo simple pero de alta fidelidad predictiva para este set de datos.